src/v/features/feature_table_snapshot.cc

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "feature_table_snapshot.h"
+
+#include "feature_table.h"
+
+namespace features {
+
+feature_table_snapshot feature_table_snapshot::from(const feature_table& ft) {
+    feature_table_snapshot fts;
+
+    fts.states.reserve(feature_schema.size());
+
+    fts.version = ft.get_active_version();
+    fts.license = ft.get_license();
+    for (const auto& state : ft._feature_state) {
+        auto& name = state.spec.name;
+        fts.states.push_back(feature_state_snapshot{
+          .name = ss::sstring(name), .state = state._state});
+    }
+
+    return fts;
+}
+
+void feature_table_snapshot::apply(feature_table& ft) const {
```

Comment:
nit: since most of this is modifying `feature_table` state, I wonder if it'd be more intuitive to make `void apply(const auto& feature_table_snapshot)` a method of the `feature_table` itself

Reply:
I was on the fence about it.  I kind of like having all the snapshotting stuff in its own file, rather than mixing it with the non-serialization related code in feature_table.  This feels like something we might revisit when doing controller snapshots in general, and come up with a uniform approach for how much the table objects should know about their own snapshot formats.